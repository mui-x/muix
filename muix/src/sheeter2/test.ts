import { compile } from './compiler2'
import { TSheeterSource, TSheeterCompiled } from './types'

const merging: TSheeterSource.RulesTreeRoot<'root' | 'disabled'> = {
    $before: {
        before: 'before',
        self: 'before',
        web: 'before',
        after: 'before',
    },
    self: 'self',
    web: 'self',
    after: 'self',
    $web: {
        web: 'web',
        after: 'web',
    },
    $after: {
        after: 'after',
    },
}

const whenUsed: TSheeterSource.RulesTreeRoot<'root' | 'disabled'> = {
    $whenUsed: {
        root: {
            plain: 'plain',
        }
    },
    ':hover': {
        $whenUsed: {
            root: {
                hover: 'hover',
            }
        },
    }
}

const whenUsedRecursion: TSheeterSource.RulesTreeRoot<'root' | 'disabled'> = {
    rule: 'rule',
    $whenUsed: {
        root: {
            plainRoot: 'plainRoot',
            $whenUsed: {
                disabled: {
                    plainInner: 'plainInner',
                    ':hover': {
                        plainInnerHover: 'plainInnerHover',
                        ':active': {
                            plainInnerHoverActive: 'plainInnerHoverActive',
                        }
                    }
                }
            },
        }
    },
    ':hover': {
        hover: 'hover',
        $whenUsed: {
            root: {
                hoverRoot: 'hoverRoot',
                $whenUsed: {
                    disabled: {
                        hoverRootInner: 'hoverRootInner',
                    }
                },
                ':active': {
                    hoverRootActive: 'hoverRootActive',
                    $whenUsed: {
                        disabled: {
                            hoverRootActiveInner: 'hoverRootActiveInner',
                        }
                    },
                },
            }
        },
    }
}



export const run = () => {
    let res
    //res = JSON.stringify(compile(merging, 'root'), null, 2)
    //res = JSON.stringify(compile(whenUsed, 'root'), null, 2)
    res = JSON.stringify(compile(whenUsedRecursion, 'root'), null, 2)
    debugger
}

