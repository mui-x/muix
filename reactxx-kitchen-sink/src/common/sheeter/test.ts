
import * as Sheeter from 'reactxx-sheeter'

export const test = () => {

  const patchable = Sheeter.toPatchableAndMergeable(root)

  const finishAddIns: Sheeter.FinishAddIns = {
    $mediaq: Sheeter.mediaqFinishAddInCreator()
  }
  const addInRulesetFilters: Sheeter.RulesetPatchGetters = {
    $mediaq: Sheeter.mediaqRulesetPatchGetterCreator(200), // screen width = 200px
    $whenUsed: Sheeter.whenUsedRulesetFilter
  }
  window.isWeb = true
  const mergedWeb = Sheeter.mergeSheetsAndFinish(patchable, [], finishAddIns)
  const codeRootWeb = Sheeter.mergeRulesetsForCode(mergedWeb, addInRulesetFilters, [mergedWeb.root, false as any, mergedWeb.b, null as any])

  window.isWeb = false
  const mergedNative = Sheeter.mergeSheetsAndFinish(patchable, [], finishAddIns)
  const codeRootNative = Sheeter.mergeRulesetsForCode(mergedNative, addInRulesetFilters, [mergedNative.root, false as any, mergedNative.b, null as any])

  debugger
}

//****************************
// INPUT DATA

const root: Sheeter.Sheet = {
  root: {
    a0: 1,
    $web: {
      a1: 1,
      ":active": {
        a2: 1,
        $before: {
          $mediaq: {
            '100-400': {
              a3: 1,
              $whenUsed: {
                b: {
                  a4: 1,
                  $after: {
                    $web: {
                      a5: 1,
                      ":hover": {
                        a6: 1,
                        $mediaq: {
                          '200-300': {
                            a7: 1,
                          }
                        } as Sheeter.Sheet
                      },
                    },
                    ":hover": {
                      $mediaq: {
                        '300-400': {
                          a8: 1,
                        }
                      } as Sheeter.Sheet
                    }
                  },
                },
              } as Sheeter.Sheet,
            },
          } as Sheeter.Sheet,
        },
      },
    },
  },
  b: {}
}

//window['isWeb'] = true

//****************************
// 

const _toPatchableAndMergeable = {
  "root": {
    "a0": 1,
    "a1": 1,
    ":active": {
      "a2": 1
    }
  },
  "b": {},
  "$addInsFinal": {
    "$mediaq": {
    }
  },
  "$addIns": {
    "$mediaq": {
      "$addIns/$whenUsed/$addIns/$mediaq/root/:active/100-200/b/:hover": {
        "300-400": {
          "a8": 1
        },
        "200-300": {
          "a7": 1
        },
        "#path": [
          "$addIns",
          "$whenUsed",
          "$addIns/$mediaq/root/:active/100-200",
          "b",
          ":hover"
        ],
      },
      "root/:active": {
        "100-200": {
          "a3": 1
        },
        "#path": [
          "root",
          ":active"
        ]
      }
    },
    "$whenUsed": {
      "$addIns/$mediaq/root/:active/100-200": {
        "b": {
          "a4": 1,
          ":hover": {
            "a6": 1
          },
          "a5": 1
        },
        "#path": [
          "$addIns",
          "$mediaq",
          "root/:active",
          "100-200"
        ]
      }
    }
  }
}

const inGetPatchLow = {
  "rootPatches": [
    {
      "path": [
        "root",
        ":active"
      ],
      "items": [
        {
          "a3": 1,
          "a4": 1,
          ":hover": {
            "a6": 1,
            "a8": 1,
            "a7": 1
          },
          "a5": 1
        }
      ]
    }
  ],
  "addInPatches": [
    {
      "path": [
        "$addIns",
        "$whenUsed",
        "$addIns/$mediaq/root/:active/100-200",
        "b",
        ":hover"
      ],
      "items": [
        {
          "a8": 1
        },
        {
          "a7": 1
        }
      ]
    },
    {
      "path": [
        "$addIns",
        "$mediaq",
        "root/:active",
        "100-200"
      ],
      "items": [
        {
          "a4": 1,
          ":hover": {
            "a6": 1,
            "a8": 1,
            "a7": 1
          },
          "a5": 1
        }
      ]
    }
  ]
}

const result = {
  "a0": 1,
  "a1": 1,
  ":active": {
    "a2": 1,
    "a3": 1,
    "a4": 1,
    ":hover": {
      "a6": 1,
      "a8": 1,
      "a7": 1
    },
    "a5": 1
  }
}