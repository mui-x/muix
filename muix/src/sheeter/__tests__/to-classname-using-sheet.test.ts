import { atomizeSheet } from "reactxx-sheeter";
import { initPlatform, toClassNames, Shape } from "reactxx-tests";
import { createSheet } from "./to-classname-sheet.test";

/*
  it("NATIVE", () => {
    initPlatform(false);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames('NATIVE: root, label, nativeOnly', null, [
      sheet.root,
      sheet.label,
      sheet.nativeOnly
    ], res => res.toMatchInlineSnapshot())

    toClassNames('NATIVE: nativeOnly, root', null, [
      sheet.nativeOnly, sheet.root
    ], res => res.toMatchInlineSnapshot())
  });
  it("WEB", () => {
    initPlatform(true);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames('WEB: root, label, webOnly', null, [
      sheet.root,
      sheet.label,
      sheet.webOnly
    ], res => res.toMatchInlineSnapshot())

    toClassNames('WEB: webOnly, root', null, [
      sheet.webOnly, sheet.root
    ], res => res.toMatchInlineSnapshot())
  });
 */

describe("TO CLASSNAMES, using sheet", () => {
  it("NATIVE", () => {
    initPlatform(false);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames(
      "NATIVE: root, label, nativeOnly",
      null,
      [sheet.root, sheet.label, sheet.nativeOnly],
      res =>
        res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  NATIVE: root, label, nativeOnly
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "margin: 4 /*root[0]*/",
    "padding: 4 /*root[0]*/",
    "padding: 10 /*root[1]/$native*/",
    "color: blue /*label*/",
    "margin: 20 /*nativeOnly/$native*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Object {
    "color": "blue /*label*/",
    "margin": "20 /*nativeOnly/$native*/",
    "padding": "10 /*root[1]/$native*/",
  },
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "color": "blue",
    "margin": 20,
    "padding": 10,
  },
]
`)
    );

    toClassNames(
      "NATIVE: nativeOnly, root",
      null,
      [sheet.nativeOnly, sheet.root],
      res =>
        res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  NATIVE: nativeOnly, root
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "margin: 20 /*nativeOnly/$native*/",
    "margin: 4 /*root[0]*/",
    "padding: 4 /*root[0]*/",
    "padding: 10 /*root[1]/$native*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Object {
    "margin": "4 /*root[0]*/",
    "padding": "10 /*root[1]/$native*/",
  },
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "margin": 4,
    "padding": 10,
  },
]
`)
    );
  });
  it("WEB", () => {
    initPlatform(true);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames(
      "WEB: root, label, webOnly",
      null,
      [sheet.root, sheet.label, sheet.webOnly],
      res =>
        res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  WEB: root, label, webOnly
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    ".a { margin:4px /*root[0]*/ }",
    ".b { padding:4px /*root[0]*/ }",
    ".c { margin:10px /*root[0]/$web*/ }",
    ".d { color:blue /*label*/ }",
    ".e { color:red /*webOnly/$web*/ }",
    ".f { padding:20px /*webOnly/$web*/ }",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Array [
    ".f { padding:20px /*webOnly/$web*/ }",
    ".e { color:red /*webOnly/$web*/ }",
    ".c { margin:10px /*root[0]/$web*/ }",
  ],
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "f e c",
]
`)
    );

    toClassNames("WEB: webOnly, root", null, [sheet.webOnly, sheet.root], res =>
      res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  WEB: webOnly, root
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    ".e { color:red /*webOnly/$web*/ }",
    ".f { padding:20px /*webOnly/$web*/ }",
    ".a { margin:4px /*root[0]*/ }",
    ".b { padding:4px /*root[0]*/ }",
    ".c { margin:10px /*root[0]/$web*/ }",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Array [
    ".c { margin:10px /*root[0]/$web*/ }",
    ".b { padding:4px /*root[0]*/ }",
    ".e { color:red /*webOnly/$web*/ }",
  ],
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "c b e",
]
`)
    );
  });
});
