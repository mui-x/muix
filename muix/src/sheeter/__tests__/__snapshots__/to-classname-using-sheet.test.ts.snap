// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TO CLASSNAMES, sheet NATIVE 1`] = `
Object {
  "label": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          "color: blue /*label*/",
        ],
      },
    ],
    "~": "c",
  },
  "nativeOnly": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          "margin: 20 /*nativeOnly/$native*/",
        ],
      },
    ],
    "~": "c",
  },
  "root": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          "margin: 4 /*root[0]*/",
          "padding: 4 /*root[0]*/",
        ],
      },
      Object {
        "atomicArray": Array [
          "padding: 10 /*root[1]/$native*/",
        ],
      },
    ],
    "~": "c",
  },
}
`;

exports[`TO CLASSNAMES, sheet WEB 1`] = `
Object {
  "label": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          ".d { color:blue /*label*/ }",
        ],
      },
    ],
    "~": "c",
  },
  "root": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          ".a { margin:4px /*root[0]*/ }",
          ".b { padding:4px /*root[0]*/ }",
        ],
      },
      Object {
        "atomicArray": Array [
          ".c { margin:10px /*root[0]/$web*/ }",
        ],
      },
    ],
    "~": "c",
  },
  "webOnly": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          ".e { color:red /*webOnly/$web*/ }",
          ".f { padding:20px /*webOnly/$web*/ }",
        ],
      },
    ],
    "~": "c",
  },
}
`;

exports[`TO CLASSNAMES, using sheet NATIVE 1`] = `
Array [
  "
##########################################
##########################################
#
#  NATIVE: root, label, nativeOnly
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "margin: 4 /*root[0]*/",
    "padding: 4 /*root[0]*/",
    "padding: 10 /*root[1]/$native*/",
    "color: blue /*label*/",
    "margin: 20 /*nativeOnly/$native*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Object {
    "color": "blue /*label*/",
    "margin": "20 /*nativeOnly/$native*/",
    "padding": "10 /*root[1]/$native*/",
  },
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "color": "blue",
    "margin": 20,
    "padding": 10,
  },
]
`;

exports[`TO CLASSNAMES, using sheet NATIVE 2`] = `
Array [
  "
##########################################
##########################################
#
#  NATIVE: nativeOnly, root
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "margin: 20 /*nativeOnly/$native*/",
    "margin: 4 /*root[0]*/",
    "padding: 4 /*root[0]*/",
    "padding: 10 /*root[1]/$native*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Object {
    "margin": "4 /*root[0]*/",
    "padding": "10 /*root[1]/$native*/",
  },
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "margin": 4,
    "padding": 10,
  },
]
`;

exports[`TO CLASSNAMES, using sheet WEB 1`] = `
Array [
  "
##########################################
##########################################
#
#  WEB: root, label, webOnly
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    ".a { margin:4px /*root[0]*/ }",
    ".b { padding:4px /*root[0]*/ }",
    ".c { margin:10px /*root[0]/$web*/ }",
    ".d { color:blue /*label*/ }",
    ".e { color:red /*webOnly/$web*/ }",
    ".f { padding:20px /*webOnly/$web*/ }",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Array [
    ".f { padding:20px /*webOnly/$web*/ }",
    ".e { color:red /*webOnly/$web*/ }",
    ".c { margin:10px /*root[0]/$web*/ }",
  ],
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "f e c",
]
`;

exports[`TO CLASSNAMES, using sheet WEB 2`] = `
Array [
  "
##########################################
##########################################
#
#  WEB: webOnly, root
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    ".e { color:red /*webOnly/$web*/ }",
    ".f { padding:20px /*webOnly/$web*/ }",
    ".a { margin:4px /*root[0]*/ }",
    ".b { padding:4px /*root[0]*/ }",
    ".c { margin:10px /*root[0]/$web*/ }",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Array [
    ".c { margin:10px /*root[0]/$web*/ }",
    ".b { padding:4px /*root[0]*/ }",
    ".e { color:red /*webOnly/$web*/ }",
  ],
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "c b e",
]
`;
